common/nexus/incoming_service_registry.go

```diff
@@ -0,0 +1,481 @@
+		DeleteService(ctx context.Context, name string) error
+		GetService(name string) (*persistencespb.VersionedNexusIncomingService, error)
+		ListServices(request *p.ListNexusIncomingServicesRequest) (*p.ListNexusIncomingServicesResponse, chan struct{}, error)
+	}
+
+	IncomingServiceRegistryConfig struct {
+		ListServicesLongPollTimeout dynamicconfig.DurationPropertyFn
+		ListServicesMinWaitTime     dynamicconfig.DurationPropertyFn
+		ListServicesRetryPolicy     backoff.RetryPolicy
+		ListServicesPageSize        dynamicconfig.IntPropertyFn
+	}
+
+	registry struct {
+		status int32
+
+		sync.RWMutex        // protects curTableVersion, servicesByName, and tableVersionChanged
+		curTableVersion     int64
+		servicesByName      map[string]*persistencespb.VersionedNexusIncomingService
+		tableVersionChanged chan struct{}
+
+		refreshPoller    *goro.Handle
+		serviceDataReady *future.FutureImpl[struct{}]
+
+		hostService primitives.ServiceName // the Temporal service that this registry instance is on
+		config      IncomingServiceRegistryConfig
+
+		matchingClient matchingservice.MatchingServiceClient
+		persistence    p.NexusServiceManager
+
+		metricsHandler metrics.Handler
+		logger         log.Logger
+	}
+)
+
+func NewIncomingServiceRegistry(
+	owningService primitives.ServiceName,
+	config IncomingServiceRegistryConfig,
+	matchingClient matchingservice.MatchingServiceClient,
+	persistence p.NexusServiceManager,
+	metricsHandler metrics.Handler,
+	logger log.Logger,
+) IncomingServiceRegistry {
+	return &registry{
+		hostService:         owningService,
+		config:              config,
+		matchingClient:      matchingClient,
+		persistence:         persistence,
+		metricsHandler:      metricsHandler,
+		logger:              logger,
+		servicesByName:      make(map[string]*persistencespb.VersionedNexusIncomingService),
+		tableVersionChanged: make(chan struct{}),
+		serviceDataReady:    future.NewFuture[struct{}](),
+	}
+}
+
+func (r *registry) Start() {
+	if !atomic.CompareAndSwapInt32(&r.status, stopped, starting) {
+		return
+	}
+	defer atomic.StoreInt32(&r.status, running)
+
+	refreshCtx := headers.SetCallerInfo(
+		context.Background(),
+		headers.SystemBackgroundCallerInfo,
+	)
+
+	if r.hostService == primitives.MatchingService {
+		r.refreshPoller = goro.NewHandle(refreshCtx).Go(r.loadServices)
+	} else {
+		r.refreshPoller = goro.NewHandle(refreshCtx).Go(r.fetchServices)
+	}
+}
+
+func (r *registry) Stop() {
+	if !atomic.CompareAndSwapInt32(&r.status, running, stopping) {
+		return
+	}
+	defer atomic.StoreInt32(&r.status, stopped)
+
+	r.refreshPoller.Cancel()
+	<-r.refreshPoller.Done()
+
+	r.persistence.Close()
+}
+
+func (r *registry) WaitUntilInitialized(ctx context.Context) error {
+	_, err := r.serviceDataReady.Get(ctx)
+	return err
+}
+
+func (r *registry) CreateOrUpdateService(ctx context.Context, service *nexus.IncomingService) (*persistencespb.VersionedNexusIncomingService, error) {
```

Comment:
If you're the owning matching node, you need to protect all of these API calls with a mutex.
We assume serial access to persistence.